### PR TITLE
[BugFix] Initialize selections in store rather than vis signals

### PR DESCRIPTION
**Pull request by djbarnwal**
Fixes Issue #4889

### CODE REVIEW

File: test/compile/selection/multi.test.ts [hunk 1]
```diff
@@ -155,9 +142,41 @@ describe('Multi Selection', () => {
...
+          }
+        ]
+      },
+      {
+        name: 'five_store',
+        values: [
+          {
+            unit: '""',
+            fields: [{type: 'E', field: 'Year'}, {type: 'E', field: 'Origin'}],
+            values: ['1970-02-02T05:00:00.000Z', 'Japan']
```
Comment (djbarnwal) [inline test/compile/selection/multi.test.ts]: Turns out Travis has a different time offset causing the checks to fail. I'll look into setting a particular timezone in jest for consistent testing across all devices.

File: src/compile/selection/assemble.ts [hunk 2]
```diff
@@ -113,9 +123,24 @@ export function assembleTopLevelSignals(model: UnitModel, signals: any[]) {
...
+        const {signals, ...rest} = proj;
+        return rest;
+      });
+      const insert = selCmpt.init.map((i: SelectionInit | SelectionInit[]) => assembleInitData(i));
+      if (selCmpt.type === 'interval') {
+        dataObj['values'] = {unit: unitName(model), fields, values: insert};
+      } else {
+        dataObj['values'] = insert.map(i => {
+          return {unit: unitName(model), fields, values: i};
+        });
```
Comment (domoritz) [inline src/compile/selection/assemble.ts]: Replace this with 

```
insert.map(i => ({unit: unitName(model), fields, values: i}));
```

**Comment by arvind** [comment]
@djbarnwal: Could you rebase this branch from master to resolve the conflicts. Thanks!

File: src/compile/selection/assemble.ts [hunk 3]
```diff
@@ -113,9 +123,22 @@ export function assembleTopLevelSignals(model: UnitModel, signals: any[]) {
...
   forEachSelection(model, selCmpt => {
+    const dataObj = {name: selCmpt.name + STORE};
+    if (selCmpt.init) {
+      const fields = selCmpt.project.items.map(proj => {
+        const {signals, ...rest} = proj;
+        return rest;
+      });
+      const insert = selCmpt.init.map((i: SelectionInit | SelectionInit[]) => assembleInitData(i));
+      if (selCmpt.type === 'interval') {
+        dataObj['values'] = {unit: unitName(model), fields, values: insert};
```
Comment (arvind) [inline src/compile/selection/assemble.ts]: Flagging this based on our in-person conversation: the square brackets notation circumvents TypeScript, so we want to use it as an absolute last resort. In this case, start by annotating `dataObj` (perhaps better called an `init`?) to be a `VgData` and then assign an array to `values`.

An additional stylistic point here -- we tend to prefer ternary assignments for simple if-then-else.

File: src/datetime.ts [hunk 4]
```diff
@@ -208,8 +208,9 @@ function normalizeDay(d: string | number) {
  * Return Vega Expression for a particular date time.
  * @param d
  * @param normalize whether to normalize quarter, month, day.
+ * @param inJSON whether to return the date in JSON format
  */
-export function dateTimeExpr(d: DateTime | DateTimeExpr, normalize = false) {
+export function dateTimeExpr(d: DateTime | DateTimeExpr, normalize = false, inJSON = false) {
```
Comment (arvind) [inline src/datetime.ts]: Let's call this parameter `toJSON`.

File: src/datetime.ts [hunk 5]
```diff
@@ -260,6 +261,16 @@ export function dateTimeExpr(d: DateTime | DateTimeExpr, normalize = false) {
...
+  if (inJSON) {
+    if (d.utc) {
+      return new Function(`return new Date(Date.UTC(${unitsString}))`)().toJSON();
+    } else {
+      return new Function(`return new Date(${unitsString})`)().toJSON();
+    }
+  }
+
   if (d.utc) {
     return `utc(${units.join(', ')})`;
```
Comment (arvind) [inline src/datetime.ts]: Since we extracted the join to a `const`, let's use it here as well.

File: package.json [hunk 6]
```diff
@@ -44,9 +44,12 @@
     "renameschema": "scripts/rename-schema.sh",
     "presite": "npm run data && npm run build:site && npm run build:toc && npm run build:versions && scripts/create-example-pages",
     "site": "bundle exec jekyll serve --incremental",
-    "prettierbase": "prettier '{src,test,test-runtime,site,typings}/**/*.{ts,md,css}'",
-    "format": "tslint -p . --fix && npm run prettierbase -- --write",
-    "lint": "tslint -p . && npm run prettierbase -- --check",
+    "tsc:src": "tsc -b src/tsconfig.src.json",
```
Comment (arvind) [inline package.json]: I'm confused about these additions to package.json scripts? They don't seem related to this PR?

File: src/compile/selection/assemble.ts [hunk 7]
```diff
@@ -123,22 +123,20 @@ export function assembleTopLevelSignals(model: UnitModel, signals: Signal[]) {
...
       });
       const insert = selCmpt.init.map((i: SelectionInit | SelectionInit[]) => assembleInitData(i));
-      if (selCmpt.type === 'interval') {
-        dataObj['values'] = {unit: unitName(model), fields, values: insert};
-      } else {
-        dataObj['values'] = insert.map(i => ({unit: unitName(model), fields, values: i}));
-      }
+      selCmpt.type === 'interval'
+        ? (init.values = [{unit: unitName(model), fields, values: insert}])
+        : (init.values = insert.map(i => ({unit: unitName(model), fields, values: i})));
```
Comment (arvind) [inline src/compile/selection/assemble.ts]: This ternary isn't structured quite right. You want it to come after the assignment:

```
init.values = selCmpt.type === 'interval' ? ... : ...;
```

File: package.json (hunk 6 above)
Comment (djbarnwal) [reply package.json]: I missed some changes while I was rebasing from the latest version of master and resolving merge conflicts. These changes just undos those changes such that `package.json` as a whole has no diff in this PR.

File: src/compile/selection/assemble.ts (hunk 7 above)
Comment (djbarnwal) [reply src/compile/selection/assemble.ts]: Oh, how silly of me 🤦‍. Correcting it right away.

**Comment by domoritz** [comment]
@djbarnwal What's the status of this PR?

**Comment by arvind** [comment]
It’s waiting for my review. I’m hoping to get to it early next week. 

-Arvind
-----
Arvind Satyanarayan <http://arvindsatya.com>

**Comment by arvind** [comment]
Nice work. Thanks @djbarnwal!

**Comment by domoritz** [comment]
It looks like master is failing after this: https://travis-ci.org/vega/vega-lite/builds/542849497

**Comment by domoritz** [comment]
Should be fixed once we merge something like https://github.com/vega/vega-lite/pull/5049

File: src/datetime.ts [hunk 8]
```diff
@@ -260,9 +261,19 @@ export function dateTimeExpr(d: DateTime | DateTimeExpr, normalize = false) {
     }
   }
 
+  const unitsString = units.join(', ');
+
+  if (toJSON) {
+    if (d.utc) {
+      return new Function(`return new Date(Date.UTC(${unitsString}))`)().toJSON();
```
Comment (domoritz) [inline src/datetime.ts]: This looks odd to me. Why are we creating a function that we execute immediately? Isn't this essentially an eval?

@arvind

File: src/datetime.ts [hunk 9]
```diff
@@ -208,8 +208,9 @@ function normalizeDay(d: string | number) {
  * Return Vega Expression for a particular date time.
  * @param d
  * @param normalize whether to normalize quarter, month, day.
+ * @param toJSON whether to return the date in JSON format
```
Comment (domoritz) [inline src/datetime.ts]: What is JSON format?
